add debugger for forked processes (#402)
inducer/pudb: doc/starting.rst

    pudb:6899: Please telnet into 127.0.0.1 6899.
    pudb:6899: Waiting for client...

Debugging forked processes
^^^^^^^^^^^^^^^^^^^^^^^^^^

On Unix system, e.g. Linux & MacOS e.g. Linux & MacOS, debugging a forked

> process is supported::

Author: When running some parts of the code in a different process, e.g. via python's multiprocessing, the tty is not attached anymore. This leads to errors for pdb and standard pudb. E.g. consider this `script.py`:
```
from multiprocessing import Process
def f(name):
    breakpoint()
    print('hello', name)

p = Process(target=f, args=('bob',))
p.start()
p.join()
```

Running `PYTHONBREAKPOINT=pudb.set_trace python script.py` results in `TypeError: ord() expected a character, but string of length 0 found` due to the missing tty. However, using `PYTHONBREAKPOINT=pudb.forked.set_trace python script.py` works in this case, since the input and output are attached explicitly. Using the `remote` pudb version would also work, but requires you to telnet into the debugger, whereas this solution directly opens the debugger on the commandline used to start python.

> process is supported::

Author: Just to highlight that this seems to be a use-case for more people:
There's a [medium article](https://medium.com/@auro_227/use-pudb-to-debug-python-multiprocessing-code-c0c5551d010c) about exactly this use-case, proposing `pudb.remote`.
Also a [stackoverflow question](https://stackoverflow.com/questions/47967490/cannot-use-pudb-for-multiprocessing), and an old issue (https://github.com/inducer/pudb/issues/286).

And even more discussion when looking for any python debugger that is able to debug forked processes, e.g. [here](https://stackoverflow.com/questions/4716533/how-to-attach-debugger-to-a-python-subproccess/23654936#23654936)

> process is supported::

Reviewer: TIL that this was a problem. That (and your general approach) make sense to me now. It would be fantastic if you could improve the docs a bit to make it a bit more obvious what problem is being solved, for example by incorporating the example you supplied. Also, I'd suggest using the subsection title "Using the debugger after forking" and describing that stdin/stdout are not usually available then.

> process is supported::

Reviewer: (And thanks for taking the time to explain!)

> process is supported::

Author: Sure thing :-) I tried to incorporate the example. Feel free to directly add suggestions if you see more room for improvement :+1:

    pudb:6899: Please telnet into 127.0.0.1 6899.
    pudb:6899: Waiting for client...

Using the debugger after forking
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In a forked process no TTY is usually attached to stdin/stdout, which leads to errors

Reviewer: ```suggestion
In a forked process, no TTY is usually attached to stdin/stdout, which leads to errors
```

when debugging with standard pudb. E.g. consider this ``script.py``::

    from multiprocessing import Process
    def f(name):
        # breakpoint was introduced in Python 3.7
        breakpoint()
        print('hello', name)
    
    p = Process(target=f, args=('bob',))
    p.start()
    p.join()

Running it with standard pudb breaks::

    PYTHONBREAKPOINT=pudb.set_trace python script.py

> However, on Unix system, e.g. Linux & MacOS, debugging a forked

Reviewer: ```suggestion
However, on Unix systems, e.g. Linux & MacOS, debugging a forked
```

    pudb:6899: Please telnet into 127.0.0.1 6899.
    pudb:6899: Waiting for client...

Using the debugger after forking
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In a forked process no tty is attached usually, which leads to
> ``TypeError: ord() expected a character, but string of length 0 found``

Reviewer: ```suggestion
In a forked process no TTY is usually attached to stdin/stdout, which leads to errors
```

when debugging with standard pudb. E.g. consider this ``script.py``::

    from multiprocessing import Process
    def f(name):
        # breakpoint was introduced in Python 3.7
        breakpoint()
        print('hello', name)
    
    p = Process(target=f, args=('bob',))
    p.start()
    p.join()

Running it with standard pudb breaks::

    PYTHONBREAKPOINT=pudb.set_trace python script.py

However, on Unix system, e.g. Linux & MacOS, debugging a forked

> process is supported using ``pudb.forked.set_trace``, which redirects
> in- and output to ``/dev/stdin`` & ``/dev/stdout``::

Reviewer: ```suggestion
process is supported using ``pudb.forked.set_trace``::
```
